File: apps/worker/app/metrics/cost.py
# Pricing table: model_name_prefix → (input_per_1m, output_per_1m) USD
PRICING: dict[str, tuple[float, float]] = {
    "gpt-4o": (5.0, 15.0),
    "gpt-4-turbo": (10.0, 30.0),
    "gpt-4": (30.0, 60.0),
    "gpt-3.5-turbo": (0.5, 1.5),
    "claude-3-5-sonnet": (3.0, 15.0),
    "claude-3-opus": (15.0, 75.0),
    "claude-3-haiku": (0.25, 1.25),
    "claude-3-sonnet": (3.0, 15.0),
    "gemini-1.5-pro": (3.5, 10.5),
    "gemini-1.5-flash": (0.35, 1.05),
}
DEFAULT_PRICING = (1.0, 3.0)


def cost_per_1m_tokens(model_name: str) -> tuple[float, float]:
    """Return (input, output) pricing per 1M tokens."""
    name_lower = model_name.lower()
    for prefix, pricing in PRICING.items():
        if name_lower.startswith(prefix):
            return pricing
    return DEFAULT_PRICING


def estimate_cost(model_name: str, input_tokens: int, output_tokens: int) -> float:
    """Return estimated cost in USD."""
    input_price, output_price = cost_per_1m_tokens(model_name)
    return (input_tokens * input_price + output_tokens * output_price) / 1_000_000

File: apps/gateway/src/lib/cost.ts
/** Input / output cost in USD per 1 M tokens */
const PRICING: Record<string, { input: number; output: number }> = {
  // OpenAI
  'gpt-4o': { input: 2.5, output: 10 },
  'gpt-4o-mini': { input: 0.15, output: 0.6 },
  'gpt-4-turbo': { input: 10, output: 30 },
  'gpt-3.5-turbo': { input: 0.5, output: 1.5 },
  // Anthropic
  'claude-3-5-sonnet-20241022': { input: 3, output: 15 },
  'claude-3-5-haiku-20241022': { input: 0.8, output: 4 },
  'claude-3-opus-20240229': { input: 15, output: 75 },
  // Mistral
  'mistral-large-latest': { input: 8, output: 24 },
  'mistral-small-latest': { input: 1, output: 3 },
  // Together / open models
  'meta-llama/Llama-3-70b-chat-hf': { input: 0.9, output: 0.9 },
  'mistralai/Mixtral-8x7B-Instruct-v0.1': { input: 0.6, output: 0.6 },
  // Groq (very cheap)
  'llama3-70b-8192': { input: 0.59, output: 0.79 },
  'mixtral-8x7b-32768': { input: 0.27, output: 0.27 },
};

const DEFAULT = { input: 1, output: 3 };

/** Returns estimated USD cost for a single LLM call */
export function estimateCostUsd(model: string, inputTokens: number, outputTokens: number): number {
  const p = PRICING[model] ?? DEFAULT;
  return (inputTokens / 1_000_000) * p.input + (outputTokens / 1_000_000) * p.output;
}
